llvm/include/llvm/Object/SYCLBIN.h

```diff
@@ -0,0 +1,86 @@
+//===- SYCLBIN.h - SYCLBIN binary format support ----------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_OBJECT_SYCLBIN_H
+#define LLVM_OBJECT_SYCLBIN_H
+
+#include "llvm/ADT/SmallString.h"
+#include "llvm/Object/Binary.h"
+#include "llvm/SYCLLowerIR/ModuleSplitter.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include <string>
+
+namespace llvm {
+
+namespace object {
+
+class SYCLBIN : public Binary {
+public:
+  SYCLBIN(MemoryBufferRef Source);
+
+  enum class BundleState : uint8_t { Input = 0, Object = 1, Executable = 2 };
+  enum class IRType : uint8_t { SPIRV = 0, PTX = 1, AMDGCN = 2 };
+
+  struct ModuleDesc {
+    BundleState State;
+    std::string ArchString;
+    std::vector<module_split::SplitModule> SplitModules;
+  };
+
+  /// The current version of the binary used for backwards compatibility.
+  static constexpr uint32_t Version = 1;
+
+  /// Magic number used to identify SYCLBIN files.
+  static constexpr uint8_t MagicNumber[4] = {0x53, 0x59, 0x42, 0x49};
+
+  /// Serialize the contents of \p ModuleDescs to a binary buffer to be read
+  /// later.
+  static Expected<SmallString<0>> write(const SmallVector<ModuleDesc> &);
+
+  static Expected<std::unique_ptr<SYCLBIN>> read(MemoryBufferRef Source);
+
+  static uint64_t getAlignment() { return 8; }
+
+  static bool classof(const Binary *V) { return V->isSYCLBINFile(); }
+
+  struct IRModule {
+    IRType Type;
+    SmallVector<char> RawIRBytes;
+  };
+  struct NativeDeviceCodeImage {
+    SmallString<0> ArchString;
+    SmallVector<char> RawDeviceCodeImageBytes;
+  };
+
+  struct AbstractModule {
+    SmallVector<SmallString<0>> KernelNames;
+    SmallVector<SmallString<0>> ImportedSymbols;
+    SmallVector<SmallString<0>> ExportedSymbols;
+    std::unique_ptr<llvm::util::PropertySetRegistry> Properties;
```

Comment:
Something that I would like to see as part of upstreaming/new offloading model transition is unification of kernel names, imported symbols and exported symbols with other properties.

From that point of view, I would suggest that we merge those into `Properties` here as well.

Reply:
They are already in there. In fact, I would like to see the opposite. I think the properties are too loose in their structure and I think having these expanded into SYCLBIN would be better.

That said, for now I agree. I will let them live in `PropertySetRegistry` for now.

Reply:
I realized that kernels are not part of properties at the moment, so they stay, but the imported and exported symbols have been removed.